doc/api/esm.md

```diff
@@ -1066,54 +1058,272 @@ const baseURL = pathToFileURL(process.cwd()).href;
+}
+```
+
+### Get source hook
+
+> Note: The loaders API is being redesigned. This hook may disappear or its
+> signature may change. Do not rely on the API described below.
+
+The `getSource` hook provides a way to define a custom method for retrieving
+the source code of an ES module specifier. This would allow a loader to
+potentially avoid reading files from disk.
+
+```js
+import { get } from 'https';
+
+/**
+ * @param {string} url
+ * @param {object} context
+ * @param {string} context.format
+ * @param {function} defaultGetSource
+ * @returns {object} response
+ * @returns {string|buffer} response.source
+ */
+export async function getSource(url, { format }, defaultGetSource) {
+  if (url.startsWith('https://')) {
+    return new Promise((resolve, reject) => {
+      get(url, (res) => {
+        let data = '';
+        res.on('data', (chunk) => data += chunk);
+        res.on('end', () => resolve({ source: data }));
+      }).on('error', (err) => reject(err));
+    });
+  }
+  return defaultGetSource(url, { format }, defaultGetSource);
+}
+
+/**
+ * @param {string} specifier
+ * @param {object} context
+ * @param {string} context.parentURL
+ * @param {function} defaultResolve
+ * @returns {object} response
+ * @returns {string} response.url
+ */
+export async function resolve(specifier,
+                              { parentURL },
+                              defaultResolve) {
+  if (specifier.startsWith('https://')) {
+    return {
+      url: specifier
+    };
+  } else if (parentURL &&
+    parentURL.startsWith('https://')) {
+    return {
+      url: new URL(specifier, parentURL).href
+    };
+  }
+  return defaultResolve({ specifier, parentURL, defaultResolve });
+}
+```
+
+With this loader, running:
+
+```console
+NODE_OPTIONS='--experimental-loader ./custom-loader.mjs' node main.js
+```
+
+Will enable `main.js` to contain a statement like
+`import { foo } from 'https://example.com/es-module.js'`, which normally isn’t
+supported by Node.js as the specifier isn’t a `file://` or `data://` URL.
+
+### Transform source hook
+
+> Note: The loaders API is being redesigned. This hook may disappear or its
+> signature may change. Do not rely on the API described below.
+
+The `transformSource` hook provides a way to modify the source code of a loaded
+ES module file after the source string has been loaded from disk but before
+Node.js has done anything with it.
+
+If this hook is used to convert unknown-to-Node.js file types into executable
+JavaScript, a resolve hook is also necessary in order to register any
+unknown-to-Node.js file extensions.
+
+```js
+import CoffeeScript from 'coffeescript';
+
+/**
+ * @param {string} source
+ * @param {object} context
+ * @param {string} context.url
+ * @param {string} context.format
+ * @param {function} defaultTransformSource
+ * @returns {object} response
+ * @returns {string|buffer} response.source
+ */
+export async function transformSource(source,
```

Comment:
`source` can be a String or Buffer here, we should note this.

Also, I'm still not completely comfortable with the fact that this is kind of up to the `getSource` implementation which exactly you get.

An alternative would be to always enforce `getSource` as returning a Buffer.